src/aws_signature_builder/mod.rs

```diff
@@ -0,0 +1,390 @@
+extern crate querystring;
+
+use crypto::digest::Digest;
+use crypto::sha2::Sha256;
+use hmac_sha256::HMAC;
+
+use hex;
+use chrono;
+
+use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
+
+use std::collections::HashMap;
+
+/// https://url.spec.whatwg.org/#fragment-percent-encode-set
+const FRAGMENT: &AsciiSet = &CONTROLS.add(b' ').add(b'"').add(b'<').add(b'>').add(b'`');
+
+/// https://url.spec.whatwg.org/#path-percent-encode-set
+const PATH: &AsciiSet = &FRAGMENT.add(b'#').add(b'?').add(b'{').add(b'}');
+
+fn normalize_query_string(query: String) -> String {
+    let mut query_pairs = querystring::querify(&query);
+    query_pairs.sort_by(|a, b| a.0.cmp(&b.0));
+    return String::from(querystring::stringify(query_pairs).trim_end_matches("&"));
+}
+
+pub fn get_utc_aws_signature_datestrings() -> (String, String) {
+    let now = chrono::Utc::now();
+    let amzdate = now.format("%Y%m%dT%H%M%SZ").to_string();
+    let datestamp = now.format("%Y%m%d").to_string();
+    return (amzdate, datestamp)
+}
+
+pub fn generate_aws_signature_headers(
+    query: String,
+    headers: HashMap<String, String>,
+    port: Option<u16>,
+    host: String,
+    amzdate: String,
+    method: String,
+    data: Vec<u8>,
+    security_token: Option<String>,
+    data_binary: bool,
+    datestamp: String,
+    service: String,
+    region: String,
+    access_key: String,
+    secret_key: String,
+    canonical_uri: String) -> HashMap<String, String> {
+    let (canonical_request,
+         payload_hash,
+         signed_headers) = task_1_create_a_canonical_request(
+         query, headers, port, host, amzdate.clone(), method, data, security_token.clone(),
+         data_binary, canonical_uri);
+    let (string_to_sign,
+         algorithm,
+         credential_scope) = task_2_create_the_string_to_sign(
+         amzdate.clone(), datestamp.clone(), canonical_request, service.clone(), region.clone());
+    let signature = task_3_calculate_the_signature(
+        datestamp, string_to_sign, service, region, secret_key);
+    let new_headers = task_4_build_auth_headers_for_the_request(
+        amzdate, payload_hash, algorithm, credential_scope, signed_headers, signature, access_key,
+        security_token);
+    return new_headers;
```

Comment:
Yeah, you calling this out makes me realize that even I don't fully understand what headers are supposed to be there and why....  I ported this from https://github.com/okigan/awscurl without touching anything since I knew that project worked for me.

[These are the docs](https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html) that I found that most closely match what that code does, so I suspect that's where it was originally copied from.

I wouldn't even be surprised if it's not even consistent across different products.  This protocol is bizarre and looks like it was designed by ten different teams.

I think I'll just paste this response into a comment actually.  Even though it's not the best, it's probably helpful to at least know that this was mainly copy-pasta.

Reply:
I added a helper function to set the headers, and an example of how to call these functions, so hopefully that makes it clearer what should be set (since they're set in one place).  Also mentioned that this code was copied.